imitation/scripts/data_collect.py

```diff
@@ -18,15 +21,91 @@ def make_PPO2(env_name, num_vec, **make_blank_policy_kwargs):
 
 
 @data_collect_ex.main
-def main(env_name, total_timesteps, num_vec=8, make_blank_policy_kwargs={}):
+def main(env_name, total_timesteps,
+         *,
+         num_vec=8,
+         make_blank_policy_kwargs={},
+
+         rollout_save: bool = False,
+         rollout_save_interval: Optional[int] = None,
+         rollout_save_n_samples: Optional[int] = None,
+         rollout_dir: Optional[str] = None,
+
+         policy_save: bool = False,
+         policy_save_interval: Optional[int] = None,
+         policy_dir: Optional[str] = None,
+         ):
   tf.logging.set_verbosity(tf.logging.INFO)
 
   policy = make_PPO2(env_name, num_vec, **make_blank_policy_kwargs)
 
-  callback = util.make_save_policy_callback("data/")
+  callback = _make_callback(
+    rollout_save, rollout_save_interval, rollout_save_n_samples,
+    rollout_dir, policy_save, policy_save_interval, policy_dir)
   policy.learn(total_timesteps, callback=callback)
 
 
+def _make_callback(rollout_save: bool = False,
+                   rollout_save_interval: Optional[int] = None,
+                   rollout_save_n_samples: Optional[int] = None,
+                   rollout_dir: Optional[str] = None,
+
+                   policy_save: bool = False,
+                   policy_save_interval: Optional[int] = None,
+                   policy_dir: Optional[str] = None,
+                   ) -> Callable:
+  """Make a callback that saves policy weights and rollouts during training.
+
+  At applicable training steps `step`,
+    - Policies are saved to `{policies_dir}/{env_name}-{step}.pkl`.
+    - Rollouts are saved to `{rollout_dir}/{env_name}-{step}.npz`.
+
+  Args:
+      rollout_save: Whether to save rollout files.
+      rollout_save_interval: The number of training updates in between
+          saving .npz rollout files.
+      rollout_save_n_samples: The number of timesteps to generate
+          and store in every .npz file.
+      rollout_dir: The directory that rollouts should be saved to.
+
+      policy_save: Whether to save policy weights.
+      policy_save_interval: The number of training updates in between
+          saving .pkl policy weight files.
+      policy_dir: The directory that policies should be saved to.
+  """
+  step = 0
+  os.makedirs(rollout_dir, exist_ok=True)
+  os.makedirs(policy_dir, exist_ok=True)
+
+  def callback(locals_: dict, _) -> bool:
+    nonlocal step
+    step += 1
+    policy = locals_['self']
+    env = policy.get_env()
+    assert env is not None
+
+    # TODO(shwang): It's kinda gross that my rollout demonstration data,
+    # which is policy independent, has a filename that depends on the policy.
+    # Maybe I should make the prefix take the policy.__class__ optionally.
+    if rollout_save and step % rollout_save_interval == 0:
```

Comment:
Since `step` is the number of times `callback` has been invoked, `rollout_save_interval` is going to have radically different meanings between RL algorithms and depending on hyperparameters. Some RL algorithms, e.g. `DQN`, call `callback` after each timestep! Others, e.g. `PPO2`, only invoke `callback` after each batch.

I solved this in https://github.com/HumanCompatibleAI/adversarial-policies/blob/master/src/modelfree/train.py by specifying a variable name and multiplier for each RL algorithm, such that the variable value in `locals_` times the multiplier equals the number of elapsed training steps. Plausibly we should patch Stable Baselines so that it always calls `callback` with some `steps_elapsed` parameter?

Reply:
I'll add an issue for this. Really liked the way your organized the experimeints in your linked file.